### DISCUSSION TITLE
Cooja long compile time

### DISCUSSION THREAD

**Discussion by rahav** [contiki-ng/contiki-ng#2113] [Q&A] [answered]
I noticed that Cooja takes much longer to compile than compiling in the container. I also suspect that it finds less problems (is that possible?). Is there a way to compile in the container to `.cooja`, then have Cooja UI just run the simulation without compiling again?

**Comment by pjonsson** [comment, marked as answer]
What OS are you using outside the container? Which version of Contiki-NG are you using? Are you talking about compiling Cooja itself, or compiling motes for a simulation?

The Cooja JAR file should be possible to move across platforms, but it takes less than 4 seconds to build on my machine:
```bash
$ time -p ant jar
...
BUILD SUCCESSFUL
Total time: 3 seconds
real 3,86
user 15,16
sys 0,41
```

If you're talking about compiling motes for a simulation on one OS and running them on something else, the answer is no, Cooja does a lot of platform dependent stuff behind the scenes.

The docker container in Contiki-NG v4.8 contains ccache though, so you should be getting a near 100% hitrate with that if you recompile the same motes repeatedly. You can see the performance improvement from ccache in the commit message for commit be26a45f7f9df5. With #2101 on latest develop, 02-compile-arm-ports takes 1m06s on a warm cache.

You can see your ccache hitrates with `ccache -s -v` and you can reset the statistics with `ccache -z`. You might get (slightly) better hitrates if you remove `-g` (debug info) from `CFLAGSNO` in `arch/platform/cooja/Makefile.cooja`.

If you are running Contiki-NG v4.8 or newer but your simulation files (`.csc`) were created with an older version of Cooja, you can enable parallel builds by editing the `<commands>` section and changing from `make <things>` to `make -j$(CPUS) <things>`. Cooja will automatically replace `$(CPUS)` with the number of cores on the machine you are running on.

**Reply by rahav** [reply]
Thank you for the detailed notes! My environment is: macOS Monterey (12.4), running VMware Fusion Professional 12.2, running Ubuntu 20.04.3 LTS.

I'm running Contiki-NG-develop/v4.7-dirty (provided with the official Contiki container).

What consumes a lot of time is Cooja compiling motes for simulation. Even on less than a handful of motes Cooja seems to compile my code as part of compiling the motes and this is what takes time.

I guess I should upgrade to Contiki 4.8 and see if its better, and again thanks for the notes. I'll mark your response as an answer until then.

**Reply by rahav** [reply]
BTW @pjonsson Do you happen to know if I can expect problems on my above research environment?

**Reply by rahav** [reply]
I wonder the performance problems I'm having is hardware related, old CPU or not enough memory? I'm using a MacBook Pro from 2016.

MacBook Pro (13-inch, 2016, Four Thunderbolt 3 Ports)
3.3 GHz Dual-Core Intel Core i7
16 GB 2133 MHz LPDDR3

I love the performance numbers you showed. May I ask you what hardware you are running on?

**Reply by rahav** [reply]
Lastly, the key reason I'm not updating to a new Mac is because I don't know whether the container version of Contiki/Cooja will run on the new Apple CPU (#1734). Predominately due to VMware, so if you happen to know about someone that runs on that would be wonderful.

**Comment by pjonsson** [comment]
The less than 4 seconds recompilation time of Cooja itself was on a Ryzen 5900x with Ubuntu 22 running in a VirtualBox on Windows 11. For compiling Cooja itself, that machine will be quite a lot faster than your laptop since it has more cores and higher boost frequencies, but you only need to compile Cooja once.

The ccache performance numbers are on the Github CI machines which have 2 cores, but beyond that I'm not sure what they are. If I had to guess, it's 2 cores of some kind of Xeon with a lot of cores, so probably 2-2.5 Ghz clock frequency. A ccache hit becomes a file copy instead of a compilation, and the CI machines seem to have bad IO performance, so the SSD in your laptop is probably not slower than the CI machines.

There are 32-bit limitations left in Cooja so it won't work with more than 2048M as maxmem parameter to the JVM. You should be fine with the remaining 14 Gb for other applications unless you do 4k video editing at the same time as your Cooja simulations.

After Cooja has compiled the motes, I mostly see 60-140% CPU usage when running make in `tests/07-simulation-base`. The highest number I saw was 204% and that was only for a second, so you're unlikely to benefit from more than 2 cores for your simulations.

I'm not aware of anyone using Cooja with an M1/M2, so you're likely to bump into early adopter problems if you go that route at this stage. You should obviously make a backup of your v4.7 tree so you can go back to it, but upgrading to v4.8 could make a significant difference for your performance issues, so I would start in that end.

**Comment by pjonsson** [comment]
Right now, `cimake` inside the docker container on 07-simulation-base takes 64 seconds with a warm ccache on my Ryzen 5900x (roughly single-core workload, so probably boosting to 4975MHz) with 3600 MHz memory. This is inside Ubuntu 22 in VirtualBox on Windows 11.

The same test on a Xeon Gold 6136 3000 MHz with 2666 MHz memory takes 91 seconds. There is no VirtualBox/Windows 11, it's just the docker container running on Ubuntu 22.

It's not my machine, but an Intel Macbook Pro that was more recent but lower clock than yours (2.3 GHz?) did the same tests in 180+ seconds.

Since August, the build system has been switched to Gradle, which is a lot faster at (re-)building, but slightly slower than the old ant way of running things (1-2s total for 26 tests when spawning a new JVM for each test). Cooja also accepts multiple simulation files as arguments now, and that reduced the runtime with 50 seconds on the Ryzen 5900x (avoided JVM restarts, perhaps some warmup of the JIT). The 32-bit limitations have been removed, but I'm guessing one has to do something pretty extreme to need/benefit from more than 2048M memory for Cooja.

Besides optimizing Cooja itself, the choice of JVM can make a difference: using OpenJDK instead of GraalVM adds roughly 3 seconds on the Ryzen 5900x, and since the JVM is only started once the 3 seconds must be more than just startup time. GraalVM is used in the docker container.

There still isn't any M1/M2 support, so the fastest machine for Cooja for now will be the top of the line gaming CPU from AMD/Intel. The Ryzen 7950x is definitely faster than a 5900x for running Cooja, and I'm guessing an Intel 13900K is even faster.

**Reply by rahav** [reply]
Much appreciated. I didn't give it any attention while I was busy with the DSN conference but I think VMware released Fusion for Apple silicon. If true, I intend to buy the biggest Apple Studio monster and give Contiki v4.8 a try. I've seen the Studio machines in (other) action and they are powerful.

**Reply by pjonsson** [reply]
For short-running simulations, you should get a significant boost from just upgrading to v4.8 on your current machine (improvements in build system so fewer things are rebuilt, and when things are rebuilt, ccache often mitigates the compile time).
